xblock/core.py

```diff
@@ -121,23 +159,530 @@ def open_local_resource(cls, uri):
+            )
+
+    def add_xml_to_node(self, node):
+        """
+        For exporting, set data on `node` from ourselves.
+        """
+        # pylint: disable=E1101
+        # Set node.tag based on our class name.
+        node.tag = self.xml_element_name()
+        node.set('xblock-family', self.entry_point)
+
+        # Set node attributes based on our fields.
+        for field_name, field in list(self.fields.items()):
+            if field_name in ('children', 'parent', 'content'):
+                continue
+            if field.is_set_on(self) or field.force_export:
+                self._add_field(node, field_name, field)
+
+        # A content field becomes text content.
+        text = self.xml_text_content()
+        if text is not None:
+            node.text = text
+
+    def xml_element_name(self):
+        """
+        What XML element name should be used for this block?
+        """
+        return self.scope_ids.block_type
+
+    def xml_text_content(self):
+        """
+        What is the text content for this block's XML node?
+        """
+        if 'content' in self.fields and self.content:  # pylint: disable=unsupported-membership-test
+            return self.content
+        else:
+            return None
+
+    def _add_field(self, node, field_name, field):
+        """
+        Add xml representation of field to node.
+
+        Depending on settings, it either stores the value of field
+        as an xml attribute or creates a separate child node.
+        """
+        value = field.to_string(field.read_from(self))
+        text_value = "" if value is None else value
+
+        # Is the field type supposed to serialize the fact that the value is None to XML?
+        save_none_as_xml_attr = field.none_to_xml and value is None
+        field_attrs = {"none": "true"} if save_none_as_xml_attr else {}
+
+        if save_none_as_xml_attr or field.xml_node:
+            # Field will be output to XML as an separate element.
+            tag = etree.QName(XML_NAMESPACES["option"], field_name)
+            elem = etree.SubElement(node, tag, field_attrs)
+            if field.xml_node:
+                # Only set the value if forced via xml_node;
+                # in all other cases, the value is None.
+                # Avoids an unnecessary XML end tag.
+                elem.text = text_value
+        else:
+            # Field will be output to XML as an attribute on the node.
+            node.set(field_name, text_value)
+
+
+class XBlockMixin(Blocklike):
     """
+    Base class for custom XBlock mixins.
+
+    To provide custom attributes to all XBlock instances in a Runtime, extend this class and
+    supply it to the Runtime's `mixins` parameter.
+
+    (...or don't do that, because it's confusing for other developers, and linters/type-checkers
+        won't understand that the mixins are part of XBlock instances, so your code will be full
+        of ignore statements :)
```

Comment:
@ormsbee You know, on second thought, what I really want is for people to avoid adding new xblock mixins to the *edx-platform* runtime. If people want to add new xblock mixins to their custom fork or custom runtime, that's A-OK. I'll move my big DO-NOT-USE-THIS comment to edx-platform instead.

Reply:
https://github.com/openedx/edx-platform/pull/34231/commits/aa8672b5a7cb784eac2fe0ec15807f4bd4c56798#diff-a16fd1c6cf75bdebdf97a2633b533d39e43975c6e98af58787ed6af1282eca71R980-R984